### PR TITLE
Fix EH issues with SPMI and crossgen2

### PR DESCRIPTION
**Pull request by davidwrighton**
There are two major adjustments here
1. Disable catching SPMI exceptions in the JIT, and instead catch them in SPMI, using a new jit interface api
2. Generalize the change @BruceForstall made in #49531 to apply to all situations where SPMI attempts to filter exceptions.

With these changes catching exceptions should be more reliable. However, I was unable to reproduce the failure, so I need some jit team expertise to identify if these fixes are good.#56050 

@BruceForstall does this solve your concerns? Could you review/test the changes to make sure they are behaving as expected?

### CODE REVIEW

**Comment by davidwrighton** [comment]
@dotnet/jit-contrib

**Comment by BruceForstall** [comment]
Does this fix https://github.com/dotnet/runtime/issues/49563?

When you write:

> However, I was unable to reproduce the failure

which failure, precisely, were you unable to reproduce?

**Comment by davidwrighton** [comment]
@BruceForstall, I tried the command for performing a superpmi collection provided in #49563, but I didn't see anything that I would identify as a problem with or without this change when I followed the instructions in that issue. So, I think it fixes the problems described in #49563, but I am unable to actually verify that. I need you to tell me if it addresses the problems you identified in that issue as I've never actually seen a superpmi collection fail.

**Comment by BruceForstall** [comment]
The testing that needs to happen is to ensure:
1. superpmi collection (esp. of crossgen2 compilation of libraries) before and after generate approximately the same number of method contexts.
2. The "catch exception" case works, but especially on Linux.

If you dump the method contexts for a collection, e.g., `c:\gh\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\mcs.exe -dump libraries.crossgen2.windows.x64.checked.mch  >dump.txt`, you can see that for some cases of GetCallInfo, the "excp" value is non-zero (this is recorded by `MethodContext::recGetCallInfo`, dumped by `MethodContext::dmpGetCallInfo`). E.g.,

`10-GetCallInfo key rt{.... lots of stuff ... excp-E06D7363`

I see 1812 such non-zero excp lines when looking at the current libraries.crossgen2.windows.x64.checked.mch file (downloaded via `superpmi.py download`). For example, the first function I see where this happens is `Microsoft.CSharp.RuntimeBinder.ComInterop.IDispatchComObject:ToString():ref:this`. (Warning: a full dump output is voluminous: about 6GB of data.)

This indicates that, during collection, the JIT called `getCallInfo`, SPMI passed it on to crossgen2, cg2 threw an exception, SPMI caught it and stored it, with the exception code, and passed it back to the JIT. On a SPMI replay of this method context, SPMI will call `RaiseException()` with this code. Will this still be caught properly by the JIT?

**Comment by davidwrighton** [comment]
I've looked at the results of running mcs.exe -dump, and it appears to be working fine. In addition to the GetCallInfo exceptions reported as 1, we're now also pulling a small number of exceptions on ResolveToken and GetMethodInfo that were not handled before.

As far as catching the exceptions, yes that still works at runtime. In those cases the crossgen2 logic isn't present. the JIT itself never actually attempts to catch the exceptions. Instead it relies on runWithErrorTrap through the JIT interface which relies on the SPMI method RunWithErrorTrap, which can catch exceptions thrown via RaiseException.

File: src/coreclr/ToolBox/superpmi/superpmi-shared/errorhandling.cpp [hunk 1]
```diff
@@ -145,3 +155,71 @@ bool RunWithErrorTrap(void (*function)(void*), void* param)
...
+        void*       pParamActual;
+    } paramStruct;
+
+    paramStruct.pParamActual = param;
+    paramStruct.function = function;
+    paramStruct.finallyFunction = finallyFunction;
+
+#ifdef HOST_UNIX
+    // We can't capture the exception code in as a PAL exceptions when thrown
+    // from crossgen2 on Linux as the jitinteface dll does not use the PAL. So
```
Comment (BruceForstall) [inline src/coreclr/ToolBox/superpmi/superpmi-shared/errorhandling.cpp:199]: >jitinteface [](http://example.com/codeflow?start=38&length=11)

typo: "jitinteface"

**Comment by BruceForstall** [comment]
> In addition to the GetCallInfo exceptions reported as 1

This is because `RunWithErrorExceptionCodeCaptureAndContinueImp` can't capture the exception thrown by crossgen2, as the comment states? That is, with your change here, we still can't catch that exception code.

**Comment by davidwrighton** [comment]
@BruceForstall, correct, the ExceptionCode can't be captured. However, as crossgen2 only through C++ exceptions, and all C++ exceptions have the same exception code, there is no particular reason to want to capture a unique code.

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**